main.py

```diff
@@ -0,0 +1,147 @@
+import threading
+
+from sorted_dict import SortedDict
+
+
+class Directions:
+    UP = 1
+    DOWN = -1
+    NONE = 0
+
+
+class HardwareElevator(object):
+    # hardware api you can just use
+    # no need to implement
+    def register_close_event_handler(self, func):
+        pass
+
+    def register_on_floor_handler(self, func):
+        pass
+
+    def move_up(self):
+        pass
+
+    def move_down(self):
+        pass
+
+    def stop_and_open_doors(self):
+        pass
+
+    def get_current_floor(self):
+        pass
+
+    def get_current_direction(self):
+        pass
+
+
+class KeepDirectionElevator(object):
+
+    def __init__(self, hardware_elevator_class=None):
+        hardware_elevator_class = hardware_elevator_class or HardwareElevator
+
+        self.hw = hardware_elevator_class()
+        self.hw.register_close_event_handler(self.on_doors_closed)
+        self.hw.register_on_floor_handler(self.on_floor)
+
+        self._drop_offs = SortedDict()
+        self._pick_ups = SortedDict()
+
+        self._direction = Directions.NONE
+        self._door_closed = False
+
+        self._start_lock = threading.Lock()
+
+    def _start(self):
+        self._clean_requests()
+
+        with self._start_lock:
+            if self._direction == Directions.NONE:
+                self._update_direction()
+
+                if self._direction == Directions.UP:
+                    self.hw.move_up()
+                elif self._direction == Directions.DOWN:
+                    self.hw.move_down()
+
+    def _stop(self, floor):
+        self._update_direction()
+
+        self._door_closed = False
+        self.hw.stop_and_open_doors()
+
+    def _update_direction(self):
+        floor = self.hw.get_current_floor()
+
+        if self._direction == Directions.NONE:
+            next_floor = self._get_next_stop(floor)
+            if next_floor > floor:
+                self._direction = Directions.UP
+                return True
+            else:
+                self._direction = Directions.DOWN
+                return True
+
+        else:
+
+            if not self._drop_offs.get_next(floor, self._direction):
+                self._direction = Directions.NONE
+                return True
+
+    def _get_next_stop(self, current_floor):
```

Comment:
In this function you are always giving priority to _drop_offs, which means that if a passenger gets into the elevator and starts clicking e.g. 4 and 5 one after another, the elevator will go between that floors and ignore other requests from outside.
This should never happen, and real elevator never give priority to the clicks inside cabin.